src/main/org/apache/tools/ant/taskdefs/optional/junitlauncher/AbstractJUnitResultFormatter.java

```diff
@@ -0,0 +1,139 @@
+        if (this.sysOutStream == null) {
+            try {
+                this.sysOutFilePath = Files.createTempFile(null, "sysout");
+                this.sysOutFilePath.toFile().deleteOnExit();
+                this.sysOutStream = Files.newOutputStream(this.sysOutFilePath);
+            } catch (IOException e) {
+                handleException(e);
+                return;
+            }
+        }
+        try {
+            this.sysOutStream.write(data);
+        } catch (IOException e) {
+            handleException(e);
+            return;
+        }
+    }
+
+    @Override
+    public void sysErrAvailable(final byte[] data) {
+        if (this.sysErrStream == null) {
+            try {
+                this.sysErrFilePath = Files.createTempFile(null, "syserr");
+                this.sysErrFilePath.toFile().deleteOnExit();
+                this.sysErrStream = Files.newOutputStream(this.sysOutFilePath);
+            } catch (IOException e) {
+                handleException(e);
+                return;
+            }
+        }
+        try {
+            this.sysErrStream.write(data);
+        } catch (IOException e) {
+            handleException(e);
+            return;
+        }
+    }
+
+    @Override
+    public void setExecutingTask(final Task task) {
+        this.task = task;
+    }
+
+    protected void writeSysOut(final Writer writer) throws IOException {
+        this.writeFrom(this.sysOutFilePath, writer);
+    }
+
+    protected void writeSysErr(final Writer writer) throws IOException {
+        this.writeFrom(this.sysErrFilePath, writer);
+    }
+
+    static Optional<TestIdentifier> traverseAndFindTestClass(final TestPlan testPlan, final TestIdentifier testIdentifier) {
+        if (isTestClass(testIdentifier).isPresent()) {
+            return Optional.of(testIdentifier);
+        }
+        final Optional<TestIdentifier> parent = testPlan.getParent(testIdentifier);
+        return parent.isPresent() ? traverseAndFindTestClass(testPlan, parent.get()) : Optional.empty();
+    }
+
+    static Optional<ClassSource> isTestClass(final TestIdentifier testIdentifier) {
+        if (testIdentifier == null) {
+            return Optional.empty();
+        }
+        final Optional<TestSource> source = testIdentifier.getSource();
+        if (!source.isPresent()) {
+            return Optional.empty();
+        }
+        final TestSource testSource = source.get();
+        if (testSource instanceof ClassSource) {
+            return Optional.of((ClassSource) testSource);
+        }
+        return Optional.empty();
+    }
+
+    private void writeFrom(final Path path, final Writer writer) throws IOException {
+        final byte[] content = new byte[1024];
+        int numBytes;
+        try (final InputStream is = Files.newInputStream(path)) {
+            while ((numBytes = is.read(content)) != -1) {
+                writer.write(new String(content, 0, numBytes));
+            }
```

Comment:
I don't think this is going to work reliably. The `read` may have split a multi-byte sequence at the end of `content` and then creating a string from it is going to break. Is there any reason you want to use a stream when reading the temporary file rather than a reader?

Reply:
You are right, good catch! I hadn't considered that scenario. I have now switched to using a reader instead.